### ISSUE TITLE
Firebase support for django - authentication, firestore and cloud functions

### ISSUE THREAD

**Issue by binoyudayan** [open-source-ideas/open-source-ideas#232]
## Project description

The idea is to use firebase cloud services for the website. As far as I know, the basic idea of firebase is server less deployment. I run into a situation where I need to use firebase and django together. I have implemented authentication using [firebase admin](https://github.com/firebase/firebase-admin-python) and [firebase-webui.](https://github.com/firebase/firebaseui-web).
At the moment I have the following things, it can be updated

1.   Deploy [authentication](https://github.com/binoyudayan/django_firebase_authetication) as an open source ( if you think its useful for somebody)
-  Login using firebase/google id token, manage the sessions
-  Handle user permissions - create user object like django auth [user](https://docs.djangoproject.com/en/3.0/topics/auth/default/#user-objects) from firebase [authentication](https://firebase.google.com/docs/auth?authuser=0) and users document(where additional user information and permissions are saved). Use this user object for different permission [mixins](https://docs.djangoproject.com/en/3.0/topics/auth/default/#limiting-access-to-logged-in-users) and context processing for user and permissions
2.  Make something similar to django Model and Manager to abstract firestore Documents and Collections(can be additional methods to existing classes in firebase admin package). Since [firestore](https://firebase.google.com/docs/firestore?authuser=0) is a non relational database, we can't define all the fields in advance. But in situations like, 
- the mandatory fields with defaults values can be defined
- document field types can be validated
- It might be useful for form field validation
- field value access using attributes instead of dictionary key
3.  Handle firebase cloud function [calls](https://firebase.google.com/docs/functions/callable-reference?authuser=0)
     - Function call, authentication, response/error handling for end user

## Relevant Technology
Python, django and firebase

### Complexity

- [ ] Beginner - This project requires no or little prior knowledge of the technolog(y|ies) specified to contribute to the project
- [x ] Intermediate - The user should have some prior knowledge of the technolog(y|ies) to the point where they know how to use it, but not necessarily all the nooks and crannies of the technology
- [ ] Advanced - The project requires the user to have a good understanding of all components of the project to contribute

### Required time (ETA)

- [ ] Little work - A couple of days
- [ ] Medium work - A week or two
- [x ] Much work - The project will take more than a couple of weeks and serious planning is required

### Categories

- [ ] Mobile app
- [ ] IoT
- [ x] Web app
- [x ] Frontend/UI
- [ ] AI/ML
- [ x] APIs/Backend
- [ ] Voice Assistant
- [ ] Developer Tooling
- [ ] Extension/Plugin/Add-On
- [ ] Design/UX
- [ ] AR/VR
- [ ] Bots
- [ x] Security
- [ ] Blockchain
- [ ] Futuristic Tech/Something Unique

**Comment by FredrikAugust** [comment]
Could you please fix the checkboxes? Thanks.

**Comment by FredrikAugust** [comment]
Also, how does this fall under "Security"? I understand that security is important to the application, but from what I gathered, it's not the focus of the idea.

**Comment by binoyudayan** [comment]
Yes, it's not the focus of the idea. I have kept 'security' as the application has to handles authentication and user sessions without any flaws. If you think, we can't keep security, I'll remove it.
